src/empirical.jl

```diff
@@ -1,68 +1,142 @@
+    sorted_values::Vector{Tuple{T, W, W, W}}
+end
+
+# type of value weights and CDF values
+weighttype(::Type{<:ECDF{<:Any, W}}) where W = W
+weighttype(ecdf::ECDF) = weighttype(typeof(ecdf))
 
-struct ECDF{T <: AbstractVector{<:Real}, W <: AbstractWeights{<:Real}}
-    sorted_values::T
-    weights::W
+# whether to interpolate between the neighboring ECDF values
+isinterpolating(::Type{<:ECDF{<:Any, <:Any, I}}) where I = I
+isinterpolating(ecdf::ECDF) = isinterpolating(typeof(ecdf))
+
+function Base.show(io::IO, e::ECDF)
+    print(io, typeof(e))
+    print(io, "(")
+    print(io, length(e.sorted_values))
+    println(io, " values)")
 end
 
+# calculate ECDF at given point
 function (ecdf::ECDF)(x::Real)
     isnan(x) && return NaN
-    n = searchsortedlast(ecdf.sorted_values, x)
-    evenweights = isempty(ecdf.weights)
-    weightsum = evenweights ? length(ecdf.sorted_values) : sum(ecdf.weights)
-    partialsum = evenweights ? n : sum(view(ecdf.weights, 1:n))
-    partialsum / weightsum
+    pos = searchsortedlast(ecdf.sorted_values, x, by=first)
+    (pos == 0) && return zero(weighttype(ecdf))
+    @inbounds val, cdf_val, val_invdelta, cdf_delta = ecdf.sorted_values[pos]
+    if !isinterpolating(ecdf) || (val == x) || (pos == length(ecdf.sorted_values))
+        return cdf_val
+    else
+        return muladd(cdf_delta, min((x - val) * val_invdelta, one(weighttype(ecdf))), cdf_val)
+    end
 end
 
-function (ecdf::ECDF)(v::RealVector)
-    evenweights = isempty(ecdf.weights)
-    weightsum = evenweights ? length(ecdf.sorted_values) : sum(ecdf.weights)
-    ord = sortperm(v)
-    m = length(v)
-    r = similar(ecdf.sorted_values, m)
-    r0 = zero(weightsum)
-    i = 1
-    for (j, x) in enumerate(ecdf.sorted_values)
-        while i <= m && x > v[ord[i]]
-            r[ord[i]] = r0
-            i += 1
-        end
-        r0 += evenweights ? 1 : ecdf.weights[j]
-        if i > m
-            break
-        end
-    end
-    while i <= m
-        r[ord[i]] = weightsum
-        i += 1
+# broadcasts ecdf() over an array
+# caches the last calculated value
+function Base.Broadcast.broadcasted(ecdf::ECDF, v::AbstractArray)
+    res = similar(v, weighttype(ecdf))
+    @inbounds for i in eachindex(v)
+        res[i] = i == 1 || v[i] != v[i-1] ? ecdf(v[i]) : res[i-1]
     end
-    return r / weightsum
+    return res
 end
 
 """
-    ecdf(X; weights::AbstractWeights)
+    ecdf(X::RealVector; weights::AbstractWeights=nothing, interpolate::Bool=false) -> ECDF
+
+Construct an *empirical cumulative distribution function* (ECDF) based on a vector of samples (`X`).
+Returns a callable [`ECDF`](@ref) object.
 
-Return an empirical cumulative distribution function (ECDF) based on a vector of samples
-given in `X`. Optionally providing `weights` returns a weighted ECDF.
+# Example
 
-Note: this function that returns a callable composite type, which can then be applied to
-evaluate CDF values on other samples.
```

Comment:
It would be nice to keep a mention about the fact that one needs to use `broadcast` here (even if that's also in examples), as it's not totally obvious.